packages/expo-device/ios/EXDevice/EXDevice.m

```diff
@@ -0,0 +1,252 @@
+// Copyright 2018-present 650 Industries. All rights reserved.
+
+#import <EXDevice/EXDevice.h>
+#import "DeviceUID.h"
+#import <UMCore/UMUtilities.h>
+
+#include <ifaddrs.h>
+#include <arpa/inet.h>
+#import <mach-o/arch.h>
+#import <CoreLocation/CoreLocation.h>
+#import <UIKit/UIKit.h>
+#import <sys/utsname.h>
+
+#import <CoreTelephony/CTCallCenter.h>
+#import <CoreTelephony/CTCall.h>
+#import <CoreTelephony/CTCarrier.h>
+#import <CoreTelephony/CTTelephonyNetworkInfo.h>
+
+#import <UMCore/UMUtilitiesInterface.h>
+#import <UMCore/UMUtilities.h>
+
+#import <WebKit/WKWebView.h>
+
+#if !(TARGET_OS_TV)
+#import <LocalAuthentication/LocalAuthentication.h>
+#endif
+
+@interface EXDevice() {
+  WKWebView *webView;
+}
+
+@property (nonatomic, strong) NSString *webViewUserAgent;
+
+@end
+
+@implementation EXDevice
+
+UM_EXPORT_MODULE(ExpoDevice);
+
+- (dispatch_queue_t)methodQueue
+{
+  return dispatch_get_main_queue();
+}
+
+- (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
+{
+  _moduleRegistry = moduleRegistry;
+}
+
+UM_EXPORT_METHOD_AS(getUserAgentAsync,
+                    getWebViewUserAgentWithResolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject)
+{
+  __weak EXDevice *weakSelf = self;
+  dispatch_async(dispatch_get_main_queue(), ^{
+    __strong EXDevice *strongSelf = weakSelf;
+    if (strongSelf) {
+      if (!strongSelf.webViewUserAgent) {
+        // We need to retain the webview because it runs an async task.
+        strongSelf->webView = [[WKWebView alloc] init];
+        
+        [strongSelf->webView evaluateJavaScript:@"window.navigator.userAgent;" completionHandler:^(id _Nullable result, NSError * _Nullable error) {
+          if (error) {
+            reject(@"ERR_CONSTANTS", error.localizedDescription, error);
+            return;
+          }
+          
+          strongSelf.webViewUserAgent = [NSString stringWithFormat:@"%@", result];
+          resolve(UMNullIfNil(strongSelf.webViewUserAgent));
+          // Destroy the webview now that it's task is complete.
+          strongSelf->webView = nil;
+        }];
+      } else {
+        resolve(UMNullIfNil(strongSelf.webViewUserAgent));
+      }
+    }
+  });
+}
+
+UM_EXPORT_METHOD_AS(getMACAddressAsync,
+                    resolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject) {
+  //some iOS privacy issues
+  NSString *address = @"02:00:00:00:00:00";
```

Comment:
@ninjaachibi Maybe cite this

Reply:
can we not get the MAC address reliably on iOS? if not, it's probably better just to not expose a method at all than to return a fake value

Reply:
I agree with Eric -- better to maybe just return null and be very clear in the documentation.

Reply:
The old approach to getting this value returns `02:00:00:00:00:00` (a cool security addition that was added) which is why this just returns the raw string.